### PR TITLE
Start the process of moving to eslint based tyepscript linting

### PR DESCRIPTION
**Pull request by lindner**
- Move main eslint config to config/eslintrc.json
- Convert js to json to help catch syntax errors
- Whitelist spacing and a few other minor issues.
- Adjust sigh to pass typescript files (and ignore gen/ directories)
- Fix sigh to check for directories explicitly
- Use throw Error instead of Promise.reject: activates rule prefer-promise-reject-errors
- Remove tabs: eslint no-tabs rule
- Fix variables declared in switch case.  Enables eslint no-case-declaration
- Fix fs import to not use require() -- Rule @typescript-eslint/no-var-requires

### CODE REVIEW

**Comment by lindner** [comment]
Adding @sjmiles since I'm touching eslint configs and @mykmartin for general reviews.

**Comment by lindner** [comment]
Addresses #3017

**Comment by sjmiles** [comment]
Can you summarize the impact on `shells/`, for which I am used to a certain eslint config. Has it changed?

**Comment by lindner** [comment]
@sjmiles  No changes to shells, all the same rules apply.  This is all about getting typescript in line with what shells has!
